hypothesis-python/src/hypothesis/strategies/_internal/strategies.py

```diff
@@ -863,14 +864,15 @@ def do_validate(self) -> None:
 
     @property
     def branches(self) -> Sequence[SearchStrategy[Ex]]:
-        if not self.__in_branches:
-            try:
-                self.__in_branches = True
-                return self.element_strategies
-            finally:
-                self.__in_branches = False
-        else:
-            return [self]
+        with self._branches_lock:
```

Comment:
Maybe we want to add a quick `if self.__element_strategies is not None: return self.element_strategies` to make this cheaper in the common case?  (or just move the lock into that part of the logic, I guess)